app/jobs/identify_external_clients_job.rb

```diff
@@ -0,0 +1,142 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# Reads objects from the inbox S3 bucket, processes each CSV file, matches clients to warehouse db, and then writes the results to the outbox S3 bucket.
+class IdentifyExternalClientsJob < BaseJob
+  # @param inbox_s3 [AwsS3] read from this bucket
+  # @param outbox_s3 [AwsS3] write into this bucket
+  def perform(inbox_s3:, outbox_s3:)
+    with_lock do
+      build_lookups
+      inbox_s3.list_objects.each do |input_object|
+        input_key = input_object.key
+        content_type = inbox_s3.get_file_type(key: input_key)
+        if content_type.present? && content_type != 'text/csv'
+          log("invalid content type #{content_type}", object_key: input_key)
+          next
+        end
+
+        data_string = inbox_s3.get_as_io(key: input_key)
+
+        input_rows = data_string ? parse_csv_string(data_string, key: input_key) : nil
+        if input_rows.blank?
+          log('invalid CSV content', object_key: input_key)
+          next
+        end
+
+        output_rows = input_rows.map { |row| process_row(row) }.compact
+        if output_rows.empty?
+          log('skipping empty output', object_key: input_key)
+          next
+        end
+
+        log("matched #{output_rows.size} of #{input_rows.size} rows", type: :info, object_key: input_key)
+
+        # s3.store raises on failure
+        outbox_s3.store(
+          content: format_as_csv(output_rows.compact),
+          name: upload_name(input_key),
+          content_type: 'text/csv; charset=utf-8',
+        )
+
+        # if we successfully processed the input object, delete it from the bucket
+        inbox_s3.delete(key: input_key)
+      end
+    end
+  end
+
+  protected
+
+  def upload_name(filename)
+    dir = File.dirname(filename)
+    dir = dir == '.' ? nil : dir
+    base = File.basename(filename, '.*')
+    ext = File.extname(filename)
+    [dir, "#{base}-results#{ext}"].compact.join('/')
+  end
+
+  private def build_lookups
+    @name_lookup = GrdaWarehouse::ClientMatcherLookups::ProperNameLookup.new(transliterate: true)
+    @ssn_lookup = GrdaWarehouse::ClientMatcherLookups::SSNLookup.new(format: :last_four)
+    @dob_lookup = GrdaWarehouse::ClientMatcherLookups::DOBLookup.new
+
+    clients = GrdaWarehouse::Hud::Client.all
```

Comment:
do we want to match all clients or just destinations?

Reply:
I believe we only want to match source clients, but return the destination ID for the matching source client.  Destinations potentially have amalgamated PII, which we probably don't want to match.